autox/watcher.go

```diff
@@ -1,48 +1,82 @@
 package autox
 
 import (
+	"fmt"
 	"os"
 
 	"github.com/iter8-tools/iter8/base/log"
 
-	// auth enables automatic authentication to various hosted clouds
-	_ "k8s.io/client-go/plugin/pkg/client/auth"
+	"helm.sh/helm/v3/pkg/cli"
 )
 
 const (
-	// Name of environment variable with file path to resource configuration yaml file
-	resourceConfigEnv = "RESOURCE_CONFIG"
-	// Name of environment variable with file path to chart group configuration yaml file
-	chartGroupConfigEnv = "CHART_GROUP_CONFIG"
+	// Name of environment variable with file path to spec group configuration yaml file
+	configEnv = "CONFIG"
 )
 
-var iter8ResourceConfig resourceConfig
-var iter8ChartGroupConfig chartGroupConfig
+var k8sClient *kubeClient
+var autoXConfig config
+
+func validateConfig(c config) error {
+	var err error
+
+	for releaseGroupSpecID, releaseGroupSpec := range autoXConfig.Specs {
+		// validate trigger
+		if releaseGroupSpec.Trigger.Namespace == "" {
+			err = fmt.Errorf("trigger in spec group \"%s\" does not have a namespace", releaseGroupSpecID)
+			break
+		}
+
+		if releaseGroupSpec.Trigger.Group == "" {
+			err = fmt.Errorf("trigger in spec group \"%s\" does not have a group", releaseGroupSpecID)
+			break
+		}
+
+		if releaseGroupSpec.Trigger.Version == "" {
+			err = fmt.Errorf("trigger in spec group \"%s\" does not have a version", releaseGroupSpecID)
+			break
+		}
+
+		if releaseGroupSpec.Trigger.Resource == "" {
+			err = fmt.Errorf("trigger in spec group \"%s\" does not have a resource", releaseGroupSpecID)
+			break
+		}
+	}
+
+	return err
+}
 
 // Start is entry point to configure services and start them
-func Start(stopCh chan struct{}) error {
+func Start(stopCh chan struct{}, autoxK *kubeClient) error {
+	if autoxK == nil {
+		// get the real client
+
+		k8sClient = newKubeClient(cli.New())
+	} else {
+		// set it here
+		k8sClient = autoxK
+	}
+
 	// initialize kubernetes driver
 	if err := k8sClient.init(); err != nil {
 		log.Logger.Fatal("unable to init k8s client")
 	}
 
-	// read resource config (resources and namespaces to watch)
-	resourceConfigFile, ok := os.LookupEnv(resourceConfigEnv)
-	if !ok {
-		log.Logger.Fatal("resource configuration file is required")
-	}
-
-	// read group config (apps and helm charts to install)
-	chartGroupConfigFile, ok := os.LookupEnv(chartGroupConfigEnv)
+	// read group config (apps and Helm charts to install)
+	configFile, ok := os.LookupEnv(configEnv)
 	if !ok {
 		log.Logger.Fatal("group configuration file is required")
 	}
 
 	// set up resource watching as defined by config
-	iter8ResourceConfig = readResourceConfig(resourceConfigFile)
-	iter8ChartGroupConfig = readChartGroupConfig(chartGroupConfigFile)
+	autoXConfig = readConfig(configFile)
+
+	err := validateConfig(autoXConfig)
+	if err != nil {
+		return err
+	}
 
-	log.Logger.Debug("chartGroupConfig:", iter8ChartGroupConfig)
+	log.Logger.Debug("config:", autoXConfig)
 
 	w := newIter8Watcher()
```

Comment:
I'm a little surprised that we aren't passing the config to the `newIter8Watcher()` ... and relying on a global variable instead. I suggest refactoring to pass the config and eliminate the global.